lib/filewriter.js

```diff
@@ -0,0 +1,182 @@
+                    self.isFirstOpen = true;
+                    return self._reopenFd(callback);
+                }
+                return callback(err);
+            }
+            self.isFirstOpen = false;
+            self.fd = fd;
+            callback(null, fd);
+        });
+    };
+
+    this._getFd = function(callback) {
+        if (this.fd !== undefined) return callback(null, this.fd);
+        else return this._reopenFd(callback);
+    };
+
+    this._getLockedFd = function(callback) {
+        var self = this;
+        this._getFd(function(err, fd) {
+            if (err) return callback(err);
+            flock(fd, "ex", function(err) {
+                if (err) return callback(err);
+                if (Date.now() < self.reopenTime) {
+                    // grab and lock the fd and only then test reopenTime, since
+                    // the expectation is that most writes will be bunched
+                    // NOTE: this is also needed for correct renameFile
+                    return callback(null, fd);
+                }
+                else {
+                    flockSync(fd, "un");
+                    self._reopenFd(function(err, fd) {
+                        if (err) return callback(err);
+                        self._getLockedFd(callback);
+                    });
+                }
+            });
+        })
+    };
+
+    // atomically append the string to the file
+    this.write = function(str, cb) {
+        var self = this;
+        this._getLockedFd(function(err, fd) {
+            if (err) return cb(err);
+            // not much benefit to reusing a buffer for the writes
+            var buf = new Buffer(str);
+            fs.write(fd, buf, 0, buf.length, null, function(err, nb) {
+                flockSync(fd, "un");
+                cb(err, nb);
+            });
+        });
+    };
+
+    // close the file
+    this.close = function( ) {
+        try { if (this.fd) fs.closeSync(this.fd); }
+        catch (err) { }
+        this.fd = undefined;
+    };
+
+    // expose class method on instance
+    this.mutexTimeout = FileWriter.mutexTimeout;
+    this.renameFile = FileWriter.renameFile;
+}
+
+
+// rename the file and wait for write activity to cease
+FileWriter.mutexTimeout = 5000;
+FileWriter.renameFile = function renameFile( oldName, newName, waitMs, cb ) {
+    if (!cb && typeof waitMs === 'function') {
+        cb = waitMs
+        waitMs = 50
+    }
+    // Wait for all pending writes to finish.  After waitMs no writer will reuse the old fd,
+    // and once we`ve paused and have the mutex we know that no more writes will occur.
+    var fd = null, mutexTimeout = (this && this.mutexTimeout) ? this.mutexTimeout : FileWriter.mutexTimeout;
+    function waitForWritesToFinish( ) {
+        var guard, error = null
+        function finish(err) {
+            if (fd !== null) fs.close(fd, function(){})
+            fd = null
+            if (guard && global.clearTimeout) clearTimeout(guard)
```

Comment:
Could there be a possible race here where fs.close initiates the call to close but guard executes before it gets to clear the timeout?  I've seen it happen before which is why I usually clear the timeout the first thing.